Source/Core/AudioCommon/Mixer.cpp

```diff
@@ -49,128 +79,154 @@ unsigned int CMixer::MixerFifo::Mix(short* samples, unsigned int numSamples, boo
-		int sampleR = ((r1 << 16) + (r2 - r1) * (u16)m_frac) >> 16;
-		sampleR = (sampleR * rvolume) >> 8;
-		sampleR += samples[currentSample];
-		samples[currentSample] = MathUtil::Clamp(sampleR, -32767, 32767);
-
-		m_frac += ratio;
-		indexR += 2 * (u16)(m_frac >> 16);
-		m_frac &= 0xffff;
+		float l_output, r_output;
+		Interpolate(read_index, &l_output, &r_output);
+		samples[current_sample + 1] += l_volume * l_output;
+		samples[current_sample] += r_volume * r_output;
+		m_fraction += ratio;
+		read_index += 2 * (s32)m_fraction;
+		m_fraction = m_fraction - (s32)m_fraction;
 	}
-
-	// Padding
-	short s[2];
-	s[0] = Common::swap16(m_buffer[(indexR - 1) & INDEX_MASK]);
-	s[1] = Common::swap16(m_buffer[(indexR - 2) & INDEX_MASK]);
-	s[0] = (s[0] * rvolume) >> 8;
-	s[1] = (s[1] * lvolume) >> 8;
-	for (; currentSample < numSamples * 2; currentSample += 2)
+	// pad output if not enough input samples
+	float s[2];
+	s[0] = m_float_buffer[(read_index - 1) & INDEX_MASK] * r_volume;
+	s[1] = m_float_buffer[(read_index - 2) & INDEX_MASK] * l_volume;
+	for (; current_sample < numSamples * 2; current_sample += 2)
 	{
-		int sampleR = MathUtil::Clamp(s[0] + samples[currentSample + 0], -32767, 32767);
-		int sampleL = MathUtil::Clamp(s[1] + samples[currentSample + 1], -32767, 32767);
-
-		samples[currentSample + 0] = sampleR;
-		samples[currentSample + 1] = sampleL;
+		samples[current_sample] += s[0];
+		samples[current_sample + 1] += s[1];
 	}
+	// update read index
+	m_read_index.store(read_index);
+}
 
-	// Flush cached variable
-	m_indexR.store(indexR);
+u32 CMixer::MixerFifo::AvailableSamples()
+{
+	return ((m_write_index.load() - m_read_index.load()) & INDEX_MASK) * 48000 / (2 * m_input_sample_rate);
+}
 
-	return numSamples;
+u32 CMixer::AvailableSamples()
+{
+	u32 samples = m_dma_mixer.AvailableSamples();
+	if (samples == 0)
+	{
+		samples = m_streaming_mixer.AvailableSamples();
+	}
+	if (samples == 0)
+	{
+		samples = m_wiimote_speaker_mixer.AvailableSamples();
+	}
+	return samples;
 }
 
-unsigned int CMixer::Mix(short* samples, unsigned int num_samples, bool consider_framelimit)
+u32 CMixer::Mix(s16* samples, u32 num_samples, bool consider_framelimit)
 {
 	if (!samples)
 		return 0;
-
-	memset(samples, 0, num_samples * 2 * sizeof(short));
-
+	std::lock_guard<std::mutex> lk(m_cs_mixing);
 	if (PowerPC::GetState() != PowerPC::CPU_RUNNING)
 	{
 		// Silence
+		memset(samples, 0, num_samples * 2 * sizeof(s16));
 		return num_samples;
 	}
+	// reset float output buffer
+	m_output_buffer.resize(num_samples * 2);
+	std::fill_n(m_output_buffer.begin(), num_samples * 2, 0.f);
+	m_dma_mixer.Mix(m_output_buffer.data(), num_samples, consider_framelimit);
+	m_streaming_mixer.Mix(m_output_buffer.data(), num_samples, consider_framelimit);
+	m_wiimote_speaker_mixer.Mix(m_output_buffer.data(), num_samples, consider_framelimit);
+	// dither and clamp
+	for (u32 i = 0; i < num_samples * 2; i += 2)
+	{
+		float r_output = m_output_buffer[i] * 38768.0f;
```

Comment:
Where does that ```38768.0f``` come from? Sounds like a magic number that should be a const or define. Or at least a comment